src/Adapter/MSTest.TestAdapter/Execution/TypeCache.cs

```diff
@@ -260,24 +260,42 @@ private static bool TryGetUnescapedManagedTypeName(TestMethod testMethod, [NotNu
     private TestClassInfo CreateClassInfo(Type classType, TestMethod testMethod)
     {
         IEnumerable<ConstructorInfo> constructors = PlatformServiceProvider.Instance.ReflectionOperations.GetDeclaredConstructors(classType);
-        ConstructorInfo? constructor = constructors.FirstOrDefault(ctor => ctor.GetParameters().Length == 0 && ctor.IsPublic);
-        bool isParameterLessConstructor;
+        (ConstructorInfo? CtorInfo, bool IsParameterless)? selectedConstructor = null;
 
-        if (classType.GetConstructor([typeof(TestContext)]) is { } testContextCtor)
+        foreach (ConstructorInfo ctor in constructors)
         {
-            constructor = testContextCtor;
-            isParameterLessConstructor = false;
-        }
-        else if (classType.GetConstructor([]) is { } parameterLessCtor)
-        {
-            constructor = parameterLessCtor;
-            isParameterLessConstructor = true;
+            if (!ctor.IsPublic)
+            {
+                continue;
+            }
+
+            ParameterInfo[] parameters = ctor.GetParameters();
+
+            // There are just 2 ctor shapes that we know, so the code is quite simple,
+            // but if we add more, add a priority to the search, and shortcircuit this search so we only iterate
+            // through the collection once, to avoid re-allocating GetParameters multiple times.
+            if (parameters.Length == 1 && parameters[0].ParameterType == typeof(TestContext))
+            {
+                selectedConstructor ??= (ctor, IsParameterless: false);
```

Comment:
If we first found a parameterless constructor, then we found this, we want to always assign right?

```suggestion
                selectedConstructor = (ctor, IsParameterless: false);
```

Reply:
I think I meant to put this on the other case, so we select the first parameterless one.